### DISCUSSION TITLE
Cross calling C

### DISCUSSION THREAD

**Discussion by samiam95124** [samiam95124/Pascal-P6#108] [General]
### What is ch2ph?

Ok, so in the series of "IP tools that would be interesting for Pascal-P6" there is is a tool called ch2ph. As some of you clever folks out there can probably guess, its a C header to  Pascal header file translator. ch2ph actually started out as a macro translator, IE., it would recognize patterns in a C language file and yield the equivalent patterns in Pascal. It actually *kinda* worked, if the "kinda" was about %90. I don't think I need to add that this is not good enough. The result was that I got serious and wrote a formal C front end (scanner and elementary parser) for ch2ph, including a full C macro preprocessor. This yeilded acceptable results, and in fact was able to translate the complete catalog of windows.h, or the windows complete API[1].

### So what good is that?

Turns out not much, since C has a number of built in irregularities. And Microsoft managed to make things far worse. Here are/were the typical cases:

1. Arrays, usually strings/arrays of character, could be either pointers or arrays.
2. Parameters could be either pointers or NULL, meaning "no parameter".
3. Parameters that could be pointers could also be small numbers so that they are clearly not pointers (both Windows and Linux treat page 0 or the first 4kb as "invalid"), leading to parameters that could be NULL, a small integer, a pointer, or an array[3].
4. String parameters could be zero terminated, or have a length, or even both at the same time.

The result of this was ch2ph used a system I'll call "annotation" where a comment with special format was prepended to parameters to tell it what to do like:

```
void cat([c_string]char*a, [c_string]char*b);
```

Then ch2ph would generate a file with instructions, in assembly, of how to treat the parameters of functions. It also took a file with general rules of thumb on how to convert things, like "char* is always zero terminated string" etc.

One fall out of ch2ph, at least with Windows, is that most C API calls get translated with the default rules. The windows.h to windows.pas translation mostly didn't need much annotation to be useful.

### So how to translate C headers to Pascal headers?

Well, first of all, there is what I'll call the "blonde translation", which means to translate the headers exactly as C specifies them. This means int becomes integer, etc. The place where this gets wonky is arrays and pointers (of course):

```
char a[] = packed array [1..maxint] of char;
char *a = ^char;
```

The places where this gets you in trouble are when you have to treat an array as a pointer and vice versa.There is also the need to pass zero terminated strings, and the fact that you are allowing C to access way off the end of any possible array.

### So how will Pascal-P6 handle calling C functions?

Well, the first rule is what I'll call the "mountain comes to Mohammed" rule. This means that you don't call C with all the quirks of C, but you place wrapper functions around C calls until it looks like something normal[2]. This is, for example, what Petit-Ami does. Its actually pretty straightforward, the most troublesome rule is with zero terminated strings, which exist in practically no other language. It turns out that they are not really very safe or portable within C itself, so for example strncat() and similar functions exist that accept non-zero terminated strings[4].

So the first answer to this question is that because Pascal-P6 obeys the calling convention of the platform it is on, you really don't have to do much. The main difference is with container strings/arrays to C string/arrays:

C:

```
void find(char a[], int len);
```

Pascaline:

```
procedure find(view s: string);
```

And thus you need annotation:

```
void find([c_lstring]char a[], int len);
```

Which tells ch2ph to convert the next two parameters to a container array. The actual binary format of the parameter is the same in %90 of the cases, which is an array base address followed by the length, although I have seen a few C functions that reverse this order.

So anything in C can be accessed from Pascaline by writing a wrapper for it. 

Now lets move on to more esoteric convertions.

### C pointer/array dualisim

So the way that most Pascal derived languages do this is via a "pointer operator" thus:

```
type ai: array [1..maxint] of integer;
        pi: ^integer;

var a: ai;
 
procedure x(p: pi);

x(@a);
```

Which finds the address of a and passes a pointer.

But this is Pascaline, a protected language. We don't drill holes in the language. Wirth himself said it. If you put a feature like this in the language, with however many warnings against its use, you will end up seeing it used everywhere.

Fortunately, Pascaline has a better and safer solution:

```
procedure x(var p: ai);
overload procedure x(p: pi);

x(a);
```

The key here is that the definitions of x(var p: ai) and x(p: pi) go to the same function in C, which takes them both as the same thing: an address of the array.

But wait, isn't this the same thing as the "find address of" operator? Well, no. In Pascaline, these two overloads are two different procedures. The uglyness of treating them as the same thing is entirely up to the called procedure, IE, C in this case. It even works for that case in Windows where the parameter could be any of NULL, a small integer, a pointer, or an array:

```
procedure x(var p: ai);
overload procedure x(p: pi);
overload procedure x(p: integer);
```
[6]

BTW this is not an ambiguous overload because numbers and variable references are distinct[5].

### What about zero terminated strings?

This is actually one case where I think a special parameter format is warranted, like:

```
procedure x(c_stringz s: string);
```

Which would convert a container string s and convert it to a zero terminated string. The reasons it is ok in the "Pascaline philosophy" are:

1. Its an action, not an interpretation. The string has to be moved to a buffer and zero placed after it.
2. Even if you use it in normal Pascaline code, there is no use for the parameter. Any use of it would be an error, and thus it is safe.

### TL;DR

The net result of this is that you end up getting a 1:1 set of Pascaline header files of .pas from the .h C files. Some of these get automatically translated, and some not. The Petit-Ami header files are hand-translated, but could be automatically converted, and at some point I may work on that. The majority of .h files can be automatically translated, but calling the APIs is not always straightforward. The majority of the cases where it is not transparent are handled via annotations. The worst cases are handled by wrappers. And yes, callbacks are some of the worst cases.

The net net is you get a "clean" Pascaline API out of the system that you can directly call. The usual alternative used in other Pascal systems is to include lots of C constructs into the language and basically let you do the work of conversions in your programs, which is both more trouble and less safe.

### Conclusion

Thus ch2ph will be a valuable tool in Pascal-P6. In fact, I have had occasional ideas about using ch2ph as the basis for a complete C to Pascaline translator.

[1] Ok, not really. Turns out that Microsoft added a lot of non-standard constructs to the C language (I know, I was shocked, shocked to find this out), so I had to accommodate that, just as I am sure I will end up including GCC quirks into the next version, I'm sure.
[2] C users will no doubt be insulted by this, so lets call it instead "the way of other languages" [than c].
[3] You see how I resisted the temptation to state how really braindead that is? Oh darn I guess I just did...
[4] The common rule used in C is that strings that are just input parameters (read only) can be zero terminated, but that result buffers returning strings get lengths. Petit-Ami uses this rule. I have considered having an alternative call for input strings that does take a length, just for calling from other languages. The reason would be that it is faster than moving the string to a buffer.
[5] Which, read the Pascaline manual, is illegal, along with chewing gum and watching Woody Allen movies.
[6] One thing I didn't try with the original ch2ph is to just generate overloads for all possible dualisim combinations by default. Its ugly, but it would mean no annotation required on these cases. An alternative is to flag externals as C externals in the compiler (something like function myfunc; cexternal) and letting the compiler treat pointers and arrays as alike just like C does. The rationale for doing that from a protected language is something like "hey, its C... its not exactly safe in any case".
